### ISSUE TITLE
InvokeDiscovery: fails after project upgrade to .NET Core 3.0

### ISSUE THREAD

**Issue by greendimka** [specsolutions/deveroom-visualstudio#60]
After I have upgraded my projects to .NET Core 3.0 Deveroom has stopped working. SpecFlow itself works - all the tests run (tested within VS2019 ver. 16.4.0).

I receive an error message within VS output window whenever I open a .feature file.

Here is the full output:

Info: CreateProjectScope: Initializing project: Core.SpecTests
Info: OnSettingsInitialized: Project settings initialized: .NETCoreApp,Version=v3.0,SpecFlow:3.1.67
Warning: InvokeDiscovery: Error during binding discovery. 
Command executed:
  C:\Users\TheUser\MyProject\Core.SpecTests\bin\Debug\netcoreapp3.0> C:\Program Files\dotnet\dotnet.exe exec C:\USERS\THEUSER\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_B4DAD456\EXTENSIONS\2NGH41Z1.JEG\Connectors\V2\deveroom-specflow-v2.dll discovery C:\Users\TheUser\MyProject\Core.SpecTests\bin\Debug\netcoreapp3.0\MyProject.SpecTests.dll ""
Exit code: 4
Message: 
Error: Exception has been thrown by the target of an invocation. -> Unable to load one or more of the requested types.
Could not load file or assembly 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.
Exception: System.Reflection.TargetInvocationException->System.Reflection.ReflectionTypeLoadException
StackTrace:
   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeModule.GetTypes()
   at System.Reflection.Assembly.GetTypes()
   at TechTalk.SpecFlow.Bindings.Discovery.RuntimeBindingRegistryBuilder.BuildBindingsFromAssembly(Assembly assembly)
   at TechTalk.SpecFlow.TestRunnerManager.BuildBindingRegistry(IEnumerable`1 bindingAssemblies)
   at TechTalk.SpecFlow.TestRunnerManager.InitializeBindingRegistry(ITestRunner testRunner)
   at TechTalk.SpecFlow.TestRunnerManager.CreateTestRunner(Int32 threadId)
   at Deveroom.VisualStudio.SpecFlowConnector.Discovery.SpecFlowV3BaseDiscoverer.GetBindingRegistry(Assembly testAssembly, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\SpecFlowV3BaseDiscoverer.cs:line 28
   at Deveroom.VisualStudio.SpecFlowConnector.Discovery.BaseDiscoverer.DiscoverInternal(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\Discovery\BaseDiscoverer.cs:line 29
   at Deveroom.VisualStudio.SpecFlowConnector.Discovery.BaseDiscoverer.Discover(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\Discovery\BaseDiscoverer.cs:line 21
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Deveroom.VisualStudio.SpecFlowConnector.ReflectionExtensions.ReflectionCallMethod[T](Object obj, String methodName, Type[] parameterTypes, Object[] args) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\ReflectionExtensions.cs:line 23
   at Deveroom.VisualStudio.SpecFlowConnector.Discovery.ReflectionSpecFlowDiscoverer.Discover(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\ReflectionSpecFlowDiscoverer.cs:line 21
   at Deveroom.VisualStudio.SpecFlowConnector.Discovery.DiscoveryProcessor.Process() in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\DiscoveryProcessor.cs:line 34
   at Deveroom.VisualStudio.SpecFlowConnector.ConsoleRunner.EntryPoint(String[] args) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\ConsoleRunner.cs:line 32

Warning: InvokeDiscovery: The project bindings (e.g. step definitions) could not be discovered. Navigation, step completion and other features are disabled. 
  Please check the error message above and report to https://github.com/specsolutions/deveroom-visualstudio/issues if you cannot fix.

**Comment by gasparnagy** [comment]
@greendimka could you please double check if the problem is not caused by some left-over files from the previous versions. E.g. close Visual Studio, try to delete the bin/obj folders, do a NuGet restore & build, and reopen Visual Studio?

If this does not help, could you please make a simple repro (an empty project with the same dependencies and a dummy feature file that shows the same error).

**Comment by greendimka** [comment]
Hi, I will do it today or tomorrow and will let you know how it went.

**Comment by greendimka** [comment]
Good news - everything does work wright in the empty project. Tried with .NET Core 3.0 and 3.1.
Bad news for me: it does not work in the mentioned project yet. I have tried everything you suggested, nothing helped, but thanks. I will keep searching for a solution.

**Comment by greendimka** [comment]
I think it is better to re-open the issue so you can close it manually if you find it appropriate.
So, I could make my project work. Have changed .NET Core from 3.0 to 3.1 - and now Deveroom is back on track again. Tried switching it back to 3.0 - got same error.

Another hint: my spec tests project references another project, which uses .NET Standard 2.1 - can it be the cause of

> Could not load file or assembly 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The system cannot find the file specified.

?

**Comment by gasparnagy** [comment]
Yes, let's keep it open. I will try to play with it a bit vased on the .NET COre 3.0/3.1 hint.

**Comment by mbendtsen** [comment]
Is this related to #61?

**Comment by JamieBenson** [comment]
I have the same issue with .Net Core 3.1 (below). Let me know what you need me to try, happy to help with legwork.

> 
> Command executed:
>   C:\TFS\Consumer Platform\Source\Main\Solutions\Tests\SeleniumGrid\SmokeTests\SeleniumSmokeTests\bin\Debug\netcoreapp3.1> C:\Program Files\dotnet\dotnet.exe exec c:\users\jbenson\appdata\local\microsoft\visualstudio\16.0_f6e6a245\extensions\ksdy0bct.cbs\Connectors\V2\deveroom-specflow-v2.dll discovery "C:\TFS\Consumer Platform\Source\Main\Solutions\Tests\SeleniumGrid\SmokeTests\SeleniumSmokeTests\bin\Debug\netcoreapp3.1\SeleniumSmokeTests.dll" ""
> Exit code: 4
> Message: 
> Error: Exception has been thrown by the target of an invocation. -> Unable to load one or more of the requested types.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
> Exception: System.Reflection.TargetInvocationException->System.Reflection.ReflectionTypeLoadException
> StackTrace:
>    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
>    at System.Reflection.RuntimeModule.GetTypes()
>    at System.Reflection.Assembly.GetTypes()
>    at TechTalk.SpecFlow.Bindings.Discovery.RuntimeBindingRegistryBuilder.BuildBindingsFromAssembly(Assembly assembly)
>    at TechTalk.SpecFlow.TestRunnerManager.BuildBindingRegistry(IEnumerable`1 bindingAssemblies)
>    at TechTalk.SpecFlow.TestRunnerManager.InitializeBindingRegistry(ITestRunner testRunner)
>    at TechTalk.SpecFlow.TestRunnerManager.CreateTestRunner(Int32 threadId)
>    at Deveroom.VisualStudio.SpecFlowConnector.Discovery.SpecFlowV3BaseDiscoverer.GetBindingRegistry(Assembly testAssembly, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\SpecFlowV3BaseDiscoverer.cs:line 28
>    at Deveroom.VisualStudio.SpecFlowConnector.Discovery.BaseDiscoverer.DiscoverInternal(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\Discovery\BaseDiscoverer.cs:line 29
>    at Deveroom.VisualStudio.SpecFlowConnector.Discovery.BaseDiscoverer.Discover(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\Discovery\BaseDiscoverer.cs:line 21
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
>    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
>    at Deveroom.VisualStudio.SpecFlowConnector.ReflectionExtensions.ReflectionCallMethod[T](Object obj, String methodName, Type[] parameterTypes, Object[] args) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\ReflectionExtensions.cs:line 23
>    at Deveroom.VisualStudio.SpecFlowConnector.Discovery.ReflectionSpecFlowDiscoverer.Discover(Assembly testAssembly, String testAssemblyPath, String configFilePath) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\ReflectionSpecFlowDiscoverer.cs:line 21
>    at Deveroom.VisualStudio.SpecFlowConnector.Discovery.DiscoveryProcessor.Process() in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector.V2\Discovery\DiscoveryProcessor.cs:line 34
>    at Deveroom.VisualStudio.SpecFlowConnector.ConsoleRunner.EntryPoint(String[] args) in W:\Deveroom\deveroom-visualstudio-prod\community\Deveroom.VisualStudio.SpecFlowConnector\ConsoleRunner.cs:line 32
> 
> Warning: InvokeDiscovery: The project bindings (e.g. step definitions) could not be discovered. Navigation, step completion and other features are disabled. 
>   Please check the error message above and report to https://github.com/specsolutions/deveroom-visualstudio/issues if you cannot fix.

**Comment by gasparnagy** [comment]
@JamieBenson To trace the issue the biggest help would be to make a minimal repro. I am using Deveroom with many netcore 3.1 projects, so this alone is not an issue, there must be something else in your setup (probably related to the dependencies). Could you try to make a small project with the same dependencies that shows the problem?

**Comment by JamieBenson** [comment]
@gasparnagy Thanks I will sort that, I was actually thinking more around are there any debugging flags I could try or whether you want to uninstall any extensions etc. :)

**Comment by gasparnagy** [comment]
@JamieBenson the discovery of the binding details is performed through an external process that is invoked by the Deveroom extension. (So installing/uninstalling the extension will not have any impact.) To try and debug, you can also run this process separately even from a command line, just copy out the command from the error message:

```
C:\TFS\Consumer Platform\Source\Main\Solutions\Tests\SeleniumGrid\SmokeTests\SeleniumSmokeTests\bin\Debug\netcoreapp3.1>"C:\Program Files\dotnet\dotnet.exe" exec c:\users\jbenson\appdata\local\microsoft\visualstudio\16.0_f6e6a245\extensions\ksdy0bct.cbs\Connectors\V2\deveroom-specflow-v2.dll discovery "C:\TFS\Consumer Platform\Source\Main\Solutions\Tests\SeleniumGrid\SmokeTests\SeleniumSmokeTests\bin\Debug\netcoreapp3.1\SeleniumSmokeTests.dll" ""
```

You can also add a `--debug` option and then the debugger will pop up. (The source code of this part is at https://github.com/specsolutions/deveroom-visualstudio/tree/master/Deveroom.VisualStudio.SpecFlowConnector. The unfortunate thing is that .NET Core is not very helpful to diagnose assembly loading issues. But it makes sense to try.

**Comment by mprochaska-landorphan** [comment]
I ran into this same issue today in netcoreapp3.1 using external binding.  It looks like discovery is working on relative to the specflow.json file before it is deployed, rather than the deployed target dll.  Copying the external dll next to the specflow.json file in the root of the project directory made discovery work.  My (sterilized) command line:

`c:\git\customer\TestProject2\bin\Debug\netcoreapp3.1> C:\Program Files\dotnet\dotnet.exe exec C:\USERS\ME\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_26D3BE02\EXTENSIONS\AO1RVX2P.FHS\Connectors\V3\deveroom-specflow-v3.dll discovery c:\git\customer\TestProject2\bin\Debug\netcoreapp3.1\TestProject2.dll c:\git\customer\TestProject2\specflow.json`

Best I can tell, discovery expects the external dll to be at the c:\git\customer\TestProject2 folder; NOT the c:\git\customer\TestProject2\bin\Debug\netcoreapp3.1 folder.

**Comment by mprochaska-landorphan** [comment]
This should not be required, but is required in the external binding projects to the target consumer projects in dotnetcore to get the tool to work:

In the external step definition project add this (this is MSBUILD in the csproj of the external binding DLL, I have one consuming project, if you have more than one consuming project, rinse and repeat):

  
  <Target Name="PostBuild" AfterTargets="PostBuildEvent" Condition="'$(SYSTEM_TEAMFOUNDATIONSERVERURI)' == ''">
    <Exec Command="copy /V /Y $(TargetPath) $(SolutionDir)**<Your Target Project Here>**\$(TargetName)$(TargetExt)" />
  </Target>

-- this copies the built external bindings DLL to the source root of the consuming test project, because that is where the discovery tool expects it to be; the discovery tool ignores the deployed directory of the consuming test project (the bug).
-- Also you must add the deployed copied DLL to .gitignore (or face merge hell)

@gasparnagy   I am in 80 hours of meetings a week; I will fill out any paper work you require when my work life slows down so that you will consider this a bug.  I think this is a SpecFlow issue, because it affects discovery of bindings.

-- I saw in a previous post you do not even have dot net core 3.1 installed; it's dated; have you fixed this?

The discovery tool looks at source, not deployed on local builds.

You closed my bug, fine, I understand.  I will never again submit an issue to you unless I can dot all I's and cross all T's; yes, I uninstalled deveroom, and will no longer recommend it to the engineers I consult with.

This is a workaround for the bug for everyone else in the world that runs into this issue when using external bindings can use.  It works on thousands of test run multiple times per day.  I am unclear whether this a a specflow issue or a deveroom issue.  I have been shut down by @gasparnagy, so I will pursue this with SpecFlow only.

I have no anger toward @gasparnagy.  If I have to have to hack MSBUILD, after 25 years in this business, I will.  I am supporting engineers with 4 months in this business.  If I have to fill out forms to get @gasparnagy to even consider his tooling is worth looking at source, not deployed, I will get around to it when I am not in crunch  (I am a consultant, nobody calls me when things are working well), I will find another tool, and stop recommending @gasparnagy's tool, which I love.

**Comment by mprochaska-landorphan** [comment]
great, the MSBUILD code was deleted

**Comment by mprochaska-landorphan** [comment]
just after build of the external bindings project, copy it to the root of the consumer projects.  Sorry the source was deleted.  Email me if you need help

**Comment by mprochaska-landorphan** [comment]
trying again:

  
  <Target Name="PostBuild" AfterTargets="PostBuildEvent" Condition="'$(SYSTEM_TEAMFOUNDATIONSERVERURI)' == ''">
    <Exec Command="copy /V /Y $(TargetPath) $(SolutionDir)Ensemble.EIQ.UI.Tests\$(TargetName)$(TargetExt)" />
  </Target>

**Comment by mprochaska-landorphan** [comment]
and again:


"
  
  <Target Name="PostBuild" AfterTargets="PostBuildEvent" Condition="'$(SYSTEM_TEAMFOUNDATIONSERVERURI)' == ''">
    <Exec Command="copy /V /Y $(TargetPath) $(SolutionDir)Ensemble.EIQ.UI.Tests\$(TargetName)$(TargetExt)" />
  </Target>

"

**Comment by mprochaska-landorphan** [comment]
Will you please contact me privately, the "safety police" will not let me post MS Build scripts.  I am certain having seen them, you will understand.

**Comment by gasparnagy** [comment]
@mprochaska-landorphan sure, please drop me a mail at gaspar at specsolutions.eu.

**Comment by gasparnagy** [comment]
There are some improvements for this in v1.6.3. Please reopen if the issue is still valid there.

**Comment by bromarock** [comment]
Hi, I've just installed version 2021.1.2.60110 on VS2022 and now I have similar problem with my project:

Info: CreateProjectScope: Initializing project: BillingNet.SpecsTests
Info: OnSettingsInitialized: Project settings initialized: .NETFramework,Version=v4.7.2,SpecFlow:2.4.0,Gen
Warning: InvokeDiscovery: Error during binding discovery. Unable to find connector: C:\USERS\DAMOMA\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\17.0_CB776178\EXTENSIONS\C2UBOIC3.T0V\V1\specflow-vs.exe
Warning: InvokeDiscovery: The project bindings (e.g. step definitions) could not be discovered. Navigation, step completion and other features are disabled. 
  Please check the error message above and report to https://github.com/SpecFlowOSS/SpecFlow.VS/issues if you cannot fix.

**Comment by gasparnagy** [comment]
@bromarock Thx for reporting this. There is a problem with the latest release. There will be a fix this afternoon. https://github.com/SpecFlowOSS/SpecFlow.VS/issues/29

**Comment by patiqs** [comment]
It seems there vas a packaging problem with the release. I just released 2021.1.4.55204 with the fix. Can you please verify?
https://marketplace.visualstudio.com/items?itemName=TechTalkSpecFlowTeam.SpecFlowForVisualStudio2022

**Comment by AleksandarMatic** [comment]
I had the same issue, it works now with this update.
